edn/src/query.rs

```diff
@@ -0,0 +1,1253 @@
+    Rc,
+};
+
+use ::{
+    BigInt,
+    DateTime,
+    OrderedFloat,
+    Uuid,
+    Utc,
+};
+
+use ::value_rc::{
+    FromRc,
+    ValueRc,
+};
+
+pub use ::{
+    Keyword,
+    PlainSymbol,
+};
+
+pub type SrcVarName = String;          // Do not include the required syntactic '$'.
+
+#[derive(Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub struct Variable(pub Rc<PlainSymbol>);
+
+impl Variable {
+    pub fn as_str(&self) -> &str {
+        self.0.as_ref().0.as_str()
+    }
+
+    pub fn to_string(&self) -> String {
+        self.0.as_ref().0.clone()
+    }
+
+    pub fn name(&self) -> PlainSymbol {
+        self.0.as_ref().clone()
+    }
+
+    /// Return a new `Variable`, assuming that the provided string is a valid name.
+    pub fn from_valid_name(name: &str) -> Variable {
+        let s = PlainSymbol::plain(name);
+        assert!(s.is_var_symbol());
+        Variable(Rc::new(s))
+    }
+}
+
+pub trait FromValue<T> {
+    fn from_value(v: &::ValueAndSpan) -> Option<T>;
+}
+
+/// If the provided EDN value is a PlainSymbol beginning with '?', return
+/// it wrapped in a Variable. If not, return None.
+/// TODO: intern strings. #398.
+impl FromValue<Variable> for Variable {
+    fn from_value(v: &::ValueAndSpan) -> Option<Variable> {
+        if let ::SpannedValue::PlainSymbol(ref s) = v.inner {
+            Variable::from_symbol(s)
+        } else {
+            None
+        }
+    }
+}
+
+impl Variable {
+    pub fn from_rc(sym: Rc<PlainSymbol>) -> Option<Variable> {
+        if sym.is_var_symbol() {
+            Some(Variable(sym.clone()))
+        } else {
+            None
+        }
+    }
+
+    /// TODO: intern strings. #398.
+    pub fn from_symbol(sym: &PlainSymbol) -> Option<Variable> {
+        if sym.is_var_symbol() {
+            Some(Variable(Rc::new(sym.clone())))
+        } else {
+            None
+        }
+    }
+}
+
+impl fmt::Debug for Variable {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "var({})", self.0)
+    }
+}
+
+impl std::fmt::Display for Variable {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub struct QueryFunction(pub PlainSymbol);
+
+impl FromValue<QueryFunction> for QueryFunction {
+    fn from_value(v: &::ValueAndSpan) -> Option<QueryFunction> {
+        if let ::SpannedValue::PlainSymbol(ref s) = v.inner {
+            QueryFunction::from_symbol(s)
+        } else {
+            None
+        }
+    }
+}
+
+impl QueryFunction {
+    pub fn from_symbol(sym: &PlainSymbol) -> Option<QueryFunction> {
+        // TODO: validate the acceptable set of function names.
+        Some(QueryFunction(sym.clone()))
+    }
+}
+
+impl std::fmt::Display for QueryFunction {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub enum Direction {
```

Comment:
We are definitely reaching the point where the `edn` crate should be renamed `parsers` or something.

Reply:
And doing so might allow us to have domain types like these as a dependency of `parsers`, rather than being _part of_ it.

Reply:
Yes, I concur.  Follow-up, either as I land (now) or as real follow-up.

Reply:
This is just a little more effort than I want to roll into this right now -- #734.